File: app/src/main/java/com/example/uscatterbrain/network/ElectLeaderPacket.java
package com.example.uscatterbrain.network;

import com.example.uscatterbrain.ScatterProto;
import com.github.davidmoten.rx2.Bytes;
import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessageLite;
import com.goterl.lazycode.lazysodium.interfaces.GenericHash;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.UUID;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

public class ElectLeaderPacket implements ScatterSerializable {
    private final ScatterProto.ElectLeader mElectLeader;
    private final byte[] salt;
    private UUID luidtag;
    public enum Phase {
        PHASE_VAL,
        PHASE_HASH
    }

    private ElectLeaderPacket(Builder builder) {
        salt = new byte[GenericHash.BYTES];
        LibsodiumInterface.getSodium().randombytes_buf(salt, salt.length);
        ScatterProto.ElectLeader.Builder b = ScatterProto.ElectLeader.newBuilder();
        if (!builder.enableHashing) {
            ScatterProto.ElectLeader.Body.Builder body = ScatterProto.ElectLeader.Body.newBuilder()
                    .setProvides(AdvertisePacket.providesToVal(builder.provides))
                    .setSalt(ByteString.copyFrom(salt));

            body.setTiebreakerVal(ScatterProto.UUID.newBuilder()
                    .setUpper(builder.tiebreaker.getMostSignificantBits())
                    .setLower(builder.tiebreaker.getLeastSignificantBits())
                    .build());


            b.setValBody(body.build());
        } else {
            b.setValHash(ByteString.copyFrom(hashFromBuilder(builder)));
        }

        mElectLeader = b.build();
    }


    private byte[] hashFromBuilder(Builder builder) {
        byte[] hashbytes = new byte[GenericHash.BYTES];
        ByteString bytes = ByteString.EMPTY;

        bytes.concat(ByteString.copyFrom(salt));
;
        bytes = bytes.concat(ByteString.copyFrom(uuidToBytes(builder.tiebreaker)));

        ByteBuffer buffer = ByteBuffer.allocate(Integer.SIZE);

        buffer.putInt(builder.provides.getVal());
        bytes.concat(ByteString.copyFrom(buffer.array()));

        LibsodiumInterface.getSodium().crypto_generichash(
                hashbytes,
                hashbytes.length,
                bytes.toByteArray(),
                bytes.toByteArray().length,
                null,
                0
        );

        return hashbytes;
    }

    public static byte[] uuidToBytes(UUID uuid) {
        ByteBuffer uuidBuffer = ByteBuffer.allocate(16);
        uuidBuffer.putLong(uuid.getMostSignificantBits());
        uuidBuffer.putLong(uuid.getLeastSignificantBits());
        return uuidBuffer.array();
    }

    public byte[] hashFromPacket() {
        byte[] hashbytes = new byte[GenericHash.BYTES];
        ByteString bytes = ByteString.EMPTY;

        bytes.concat(ByteString.copyFrom(salt));

        bytes = bytes.concat(ByteString.copyFrom(uuidToBytes(getTieBreak())));

        ByteBuffer buffer = ByteBuffer.allocate(Integer.SIZE);

        buffer.putInt(mElectLeader.getValBody().getProvides());
        bytes.concat(ByteString.copyFrom(buffer.array()));

        LibsodiumInterface.getSodium().crypto_generichash(
                hashbytes,
                hashbytes.length,
                bytes.toByteArray(),
                bytes.toByteArray().length,
                null,
                0
        );

        return hashbytes;
    }

    public boolean verifyHash(ElectLeaderPacket packet) {
        if (packet.isHashed() == this.isHashed()) {
            return false;
        } else if (this.isHashed()) {
            byte[] hash =  packet.hashFromPacket();
            return Arrays.equals(hash, getHash());
        } else {
            byte[] hash = hashFromPacket();
            return Arrays.equals(hash, packet.getHash());
        }
    }

    private ElectLeaderPacket(InputStream inputStream) throws IOException {
        this.mElectLeader = ScatterProto.ElectLeader.parseDelimitedFrom(inputStream);
        if (this.mElectLeader.getValCase().compareTo(ScatterProto.ElectLeader.ValCase.VAL_BODY) == 0) {
            this.salt = mElectLeader.getValBody().getSalt().toByteArray();
        } else {
            this.salt = new byte[0];
        }
    }

    public UUID getTieBreak() {
        return new UUID(
                mElectLeader.getValBody().getTiebreakerVal().getUpper(),
                mElectLeader.getValBody().getTiebreakerVal().getLower()
        );
    }

    @Override
    public void tagLuid(UUID luid) {
        luidtag = luid;
    }

    @Override
    public UUID getLuid() {
        return luidtag;
    }

    public AdvertisePacket.Provides getProvides() {
        return AdvertisePacket.valToProvides(mElectLeader.getValBody().getProvides());
    }

    public static Single<ElectLeaderPacket> parseFrom(InputStream inputStream) {
        return Single.fromCallable(() -> new ElectLeaderPacket(inputStream));
    }

    public static Single<ElectLeaderPacket> parseFrom(Observable<byte[]> flowable) {
        InputStreamObserver observer = new InputStreamObserver();
        flowable.subscribe(observer);
        return ElectLeaderPacket.parseFrom(observer).doFinally(observer::close);
    }

    public static Single<ElectLeaderPacket> parseFrom(Flowable<byte[]> flowable) {
        InputStreamFlowableSubscriber observer = new InputStreamFlowableSubscriber();
        flowable.subscribe(observer);
        return ElectLeaderPacket.parseFrom(observer).doFinally(observer::close);
    }

    @Override
    public byte[] getBytes() {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            mElectLeader.writeDelimitedTo(os);
            return os.toByteArray();
        } catch (IOException ignored) {
            return null;
        }
    }

    @Override
    public ByteString getByteString() {
        return ByteString.copyFrom(getBytes());
    }

    @Override
    public Completable writeToStream(OutputStream os) {
        return Completable.fromAction(() -> mElectLeader.writeDelimitedTo(os));
    }

    @Override
    public GeneratedMessageLite getMessage() {
        return mElectLeader;
    }

    @Override
    public Flowable<byte[]> writeToStream(int fragize) {
        return Bytes.from(new ByteArrayInputStream(getBytes()), fragize);
    }

    @Override
    public PacketType getType() {
        return PacketType.TYPE_ELECT_LEADER;
    }

    public boolean isHashed() {
        return mElectLeader.getValCase() == ScatterProto.ElectLeader.ValCase.VAL_HASH;
    }

    public byte[] getHash() {
        return mElectLeader.getValHash().toByteArray();
    }

    public static Builder newBuilder() {
        return new Builder();
    }

    public static class Builder {

        private boolean enableHashing = false;
        private AdvertisePacket.Provides provides;
        private UUID tiebreaker;

        private Builder() {
        }

        public Builder enableHashing() {
            this.enableHashing = true;
            return this;
        }

        public Builder setProvides(AdvertisePacket.Provides provides) {
            this.provides = provides;
            return this;
        }


        public Builder setTiebreaker(UUID tiebreaker) {
            this.tiebreaker =tiebreaker;
            return this;
        }

        public ElectLeaderPacket build() {
            if ( provides == null || tiebreaker == null) {
                throw new IllegalArgumentException("both tiebreaker and provides must be set");
            }

            return new ElectLeaderPacket(this);
        }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/proto/ElectLeaderPacket.kt
package net.ballmerlabs.uscatterbrain.network.proto

import com.google.protobuf.ByteString
import com.goterl.lazysodium.interfaces.GenericHash
import net.ballmerlabs.sbproto.SbPacket
import proto.Scatterbrain.ElectLeader

import java.nio.ByteBuffer
import java.util.UUID
import net.ballmerlabs.scatterproto.*
import net.ballmerlabs.uscatterbrain.network.LibsodiumInterface
import proto.Scatterbrain.MessageType

/**
 * wrapper class for ElectLeader protobuf message
 *
 * @property tieBreak tiebreak value
 * @property provides AdvertisePacket provides value
 * @property isHashed true if this packet only contains a hash value
 * @property hash hash value if this packet is hashed
 */
@SbPacket(messageType = MessageType.ELECT_LEADER)
class ElectLeaderPacket(
    packet: ElectLeader
) : ScatterSerializable<ElectLeader>(packet, MessageType.ELECT_LEADER) {
    private fun hashFromPacket(): ByteArray {
        return if (isHashed) {
            packet.valHash.toByteArray()
        } else {
            val hashbytes = ByteArray(GenericHash.BYTES)
            var bytes = ByteString.EMPTY
            bytes.concat(packet.valBody.salt)
            bytes = bytes.concat(ByteString.copyFrom(
                ElectLeaderPacket.Companion.uuidToBytes(
                    tieBreak
                )
            ))
            val buffer = ByteBuffer.allocate(Integer.SIZE)
            buffer.putInt(packet.valBody.provides)
            bytes.concat(ByteString.copyFrom(buffer.array()))
            LibsodiumInterface.sodium.crypto_generichash(
                hashbytes,
                hashbytes.size,
                bytes.toByteArray(),
                bytes.toByteArray().size.toLong(),
                null,
                0
            )
            hashbytes
        }
    }

    fun verifyHash(remote: ElectLeaderPacket): Boolean {
        return when {
            remote.isHashed == isHashed -> {
                false
            }

            isHashed -> {
                val hash = remote.hashFromPacket()
                hash.contentEquals(hash)
            }

            else -> {
                val hash = hashFromPacket()
                hash.contentEquals(remote.hash)
            }
        }
    }

    val from: UUID
        get() = packet.sender.toUuid()
    val tieBreak: UUID
        get() = UUID(
            packet.valBody.tiebreakerVal.upper, packet.valBody.tiebreakerVal.lower
        )

    val upgrade: Optional<UpgradePacket>
        get() = if (packet.valBody.maybeUpgradeCase == ElectLeader.Body.MaybeUpgradeCase.UPGRADE) {
            Optional.of(UpgradePacket(packet.valBody.upgrade))
        } else {
            Optional.empty()
        }

    val provides: Provides
        get() = AdvertisePacket.Companion.valToProvides(
            packet.valBody.provides
        )

    val isHashed: Boolean
        get() = packet.valCase == ElectLeader.ValCase.VAL_HASH

    val hash: ByteArray
        get() = packet.valHash.toByteArray()

    override fun validate(): Boolean {
        return hash.size <= MAX_HASH
    }

    data class Builder(
        val sender: UUID?,
        var enableHashing: Boolean = false,
        var hashVal: ByteString? = null,
        var provides: Provides? = null,
        var tiebreaker: UUID? = null,
        var upgrade: UpgradePacket? = null
    ) {
        private val salt: ByteArray = ByteArray(GenericHash.BYTES)


        private fun hashFromBuilder(): ByteString {
            val hashbytes = ByteArray(GenericHash.BYTES)
            var bytes = ByteString.EMPTY
            bytes.concat(ByteString.copyFrom(salt))
            bytes = bytes.concat(ByteString.copyFrom(
                ElectLeaderPacket.Companion.uuidToBytes(
                    tiebreaker!!
                )
            ))
            val buffer = ByteBuffer.allocate(Integer.SIZE)
            buffer.putInt(provides!!.`val`)
            bytes.concat(ByteString.copyFrom(buffer.array()))
            LibsodiumInterface.sodium.crypto_generichash(
                hashbytes,
                hashbytes.size,
                bytes.toByteArray(),
                bytes.toByteArray().size.toLong(),
                null,
                0
            )
            return ByteString.copyFrom(hashbytes)
        }

        fun enableHashing() = apply {
            enableHashing = true
        }

        fun setProvides(provides: Provides) = apply {
            this.provides = provides
        }

        fun setTiebreaker(tiebreaker: UUID) = apply {
            this.tiebreaker = tiebreaker
        }

        fun setHash(hash: ByteString) = apply {
            this.hashVal = hash
        }

        fun setUpgrade(upgrade: UpgradePacket) = apply {
            this.upgrade = upgrade
        }

        fun build(): ElectLeaderPacket {
            require((!(provides == null || tiebreaker == null)) || hashVal != null) { "both tiebreaker and provides must be set" }
            return if (enableHashing) {
                ElectLeaderPacket(
                    ElectLeader.newBuilder().setValHash(hashFromBuilder())
                        //.setType(ScatterProto.MessageType.ELECT_LEADER)
                        .build()
                )
            } else {
                val builder = ElectLeader.Body.newBuilder().setSalt(ByteString.copyFrom(salt))
                    .setProvides(
                        providesToVal(
                            provides!!
                        )
                    ).setTiebreakerVal(tiebreaker?.toProto())
                val u = upgrade
                if (u != null) {
                    builder.setUpgrade(u.packet)
                }
                ElectLeaderPacket(
                    ElectLeader.newBuilder().setSender(sender?.toProto()).setValBody(
                        builder.build()
                    )
                        //.setType(ScatterProto.MessageType.ELECT_LEADER)
                        .build()
                )
            }
        }
    }

    companion object {
        fun uuidToBytes(uuid: UUID): ByteArray {
            val uuidBuffer = ByteBuffer.allocate(16)
            uuidBuffer.putLong(uuid.mostSignificantBits)
            uuidBuffer.putLong(uuid.leastSignificantBits)
            return uuidBuffer.array()
        }

        fun newBuilder(sender: UUID? = null): Builder {
            return Builder(sender)
        }
    }
}